File: src/main/java/com/ada/conversation/context/ContextAssembler.java
package com.ada.conversation.context;

import com.ada.conversation.application.dto.*;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.core.annotation.AnnotationAwareOrderComparator;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
public class ContextAssembler {
  private final List<ContextItem> items;

  public ContextState build(ChatRequest r) {
    var current = new ContextState();
    var ordered = new java.util.ArrayList<>(items);
    ordered.sort(AnnotationAwareOrderComparator.INSTANCE);
    for (var item : ordered) current = item.apply(r, current);
    return current;
  }
}

File: src/main/kotlin/com/ada/conversation/context/ContextAssembler.kt
package com.ada.conversation.context

import com.ada.conversation.application.dto.ChatRequest
import com.ada.conversation.application.dto.ContextContribution
import org.springframework.core.annotation.AnnotationAwareOrderComparator
import org.springframework.stereotype.Component

@Component
class ContextAssembler(
    private val contextItems: List<ContextItem>,
) {
    fun build(request: ChatRequest): ContextContribution = orderedItems()
        .map { it.build(request) }
        .fold(ContextContribution()) { result, contribution ->
            ContextContribution(
                messages = result.messages + contribution.messages,
                tools = result.tools + contribution.tools,
            )
        }

    private fun orderedItems(): List<ContextItem> = contextItems.sortedWith(AnnotationAwareOrderComparator.INSTANCE)
}
